talk/morelanguage/exceptions.tex

```diff
@@ -4,185 +4,166 @@
   \frametitlecpp[98]{Exceptions}
   \begin{block}{The concept}
     \begin{itemize}
-    \item exceptional event breaking linearity of the code
-    \item will be handled in dedicated place
+      \item to handle \textit{exceptional} events that cannot be foreseen
+      \item to cleanly jump to the place where the error can be handled
     \end{itemize}
   \end{block}
-  \begin{block}{Practically}
+  \begin{block}{In practice}
     \begin{itemize}
-    \item you can throw any object with {\it throw}
-    \item you handle them using {\it try ... catch} blocks
+      \item add an exception handling block with {\it try ... catch}
+      \item throw an exception within it using {\it throw}
     \end{itemize}
   \end{block}
-  \begin{cppcode}
+  \begin{cppcode*}{fontsize=\scriptsize}
+    #include <stdexcept>
+    ...
     try {
-      if (0 == name) {
-        throw std::string("Expected non empty name");
+      if (0 == divisor) {
+        throw illegal_argument("Cannot divide by zero");
       }
-      printf("%s\n", name);
-    } catch (std::string& e) {
-      printf("empty name found\n");
+      ... // Normal processing
+    } catch (const invalid_argument& e) {
+      cerr << e.what() << endl;
     }
-  \end{cppcode}
+  \end{cppcode*}
 \end{frame}
 
 \begin{frame}[fragile]
   \frametitlecpp[98]{Exceptions}
-  \begin{block}{Rules}
+  \begin{block}{Rules and Advice}
     \begin{itemize}
-    \item exception will skip all code until next {\it catch}
-    \begin{itemize}
-      \item still destructors are called when exiting scopes
-      \item but your own cleanup may not be
+      \item any object can be thrown, but best to use those in \texttt{stdexcept}
+      \begin{itemize}
+        \item define your own subclass of \texttt{std::exception} if needed
+      \end{itemize}
+      \item exceptions will be caught if the object type of \textit{catch} matches
+      \begin{itemize}
+          \item if nothing catches an exception then \texttt{std::terminate} is called
+          \item throw exceptions by value, catch them by (const) reference
+      \end{itemize}
+      \item all objects on the stack between the \textit{throw} and the \textit{catch} have their destructors called
+      \begin{itemize}
+        \item this should give you a clean release of intermediate resources
+        \item make sure you are using the RAII idiom in your own classes
+      \end{itemize}
+      \item use exceptions for runtime errors like bad inputs, file not found, DB connection, \ldots
+      \item \textit{don't} use exceptions for logic errors in your code (use \texttt{assert})
```

Comment:
`assert` are useless in release builds, so you either get wrong results or a crash.
And I think there's logic errors that would deserve some kind of clean termination. At least an exception has a `what()` message that you could inspect.
You can also print an error message and `abort`, so I think we should find some general advice we can give to users.

Reply:
So a logic error means that you made a programming mistake so this is, by definiton, unrecovereable - the program needs fixed. All an exception does here is beautify things. It's also doubtful here that `e.what()` can tell you anything useful and you now lost the stack that could have been useful (I think even if you re-throw you lost some context?).

I still think `assert()` is the useful way to test all of the possible outcomes, although it's sometimes hard to anticipate all cases, that's true...

Reply:
With `consider assert and tests`, you went in the right direction. I wanted to prevent people from writing `assert(false)`, which I have seen too often, but that's off the table now! 👍